File: backend/app/api/integrations.py
from fastapi import APIRouter, Depends, Query, HTTPException
from sqlalchemy.orm import Session
from app.database.session import get_db
from app.auth.dependencies import get_current_user
from app.models.user import User
from app.schemas.integration import IntegrationResponse
from app.schemas.account import AccountResponse
from app.services.integration_service import IntegrationService
from app.integrations.registry.integration_registry import IntegrationRegistry
from app.auth.jwt import generate_oauth_state, verify_oauth_state

router = APIRouter(
    prefix="/integrations",
    tags=["Integrations"]
)

@router.get("", response_model=list[IntegrationResponse])
def get_integrations(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return IntegrationService.get_user_integrations(db, current_user.id)

@router.get("/accounts", response_model=list[AccountResponse])
def get_accounts(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return IntegrationService.get_user_accounts(db, current_user.id)

@router.post("/connect", response_model=IntegrationResponse)
async def connect_integration(
    provider: str,
    code: str,
    redirect_uri: str,
    state: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    if not verify_oauth_state(state, current_user.id, provider):
        raise HTTPException(status_code=400, detail="Invalid OAuth state parameter")
        
    return await IntegrationService.connect_integration(
        db, current_user.id, provider, code, redirect_uri
    )

@router.get("/auth-url")
async def get_auth_url(
    provider: str,
    redirect_uri: str,
    current_user: User = Depends(get_current_user)
):
    try:
        connector = IntegrationRegistry.get_connector(provider)
        state = generate_oauth_state(current_user.id, provider)
        url = await connector.get_authorization_url(redirect_uri, state)
        return {"url": url, "state": state}
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.delete("/{integration_id}")
def disconnect_integration(
    integration_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    IntegrationService.disconnect_integration(db, current_user.id, integration_id)
    return {"message": "Disconnected successfully"}

File: frontend/services/integrations.ts
import apiClient from "./api";

export interface Integration {
  id: number;
  user_id: number;
  provider: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface Account {
  id: number;
  user_id: number;
  integration_id: number;
  platform_id: string;
  name: string;
  profile_picture?: string;
  is_active: boolean;
  metadata_json?: {
    platform: string;
    username?: string;
    category?: string;
    facebook_page_id?: string;
  };
  created_at: string;
  updated_at: string;
}

export const integrationsService = {
  async getIntegrations(): Promise<Integration[]> {
    const response = await apiClient.get<Integration[]>("/integrations");
    return response.data;
  },

  async getAccounts(): Promise<Account[]> {
    const response = await apiClient.get<Account[]>("/integrations/accounts");
    return response.data;
  },

  async getAuthUrl(provider: string, redirectUri: string): Promise<{ url: string; state: string }> {
    const response = await apiClient.get<{ url: string; state: string }>("/integrations/auth-url", {
      params: { provider, redirect_uri: redirectUri },
    });
    return response.data;
  },

  async connectIntegration(provider: string, code: string, redirectUri: string, state: string): Promise<Integration> {
    const response = await apiClient.post<Integration>("/integrations/connect", null, {
      params: { provider, code, redirect_uri: redirectUri, state },
    });
    return response.data;
  },

  async disconnectIntegration(integrationId: number): Promise<void> {
    await apiClient.delete(`/integrations/${integrationId}`);
  },
};
